x-pack/filebeat/input/awss3/sqs_s3_event.go

```diff
@@ -0,0 +1,255 @@
+	// after maximum receives is reached.
+	p.metrics.sqsMessagesReturnedTotal.Inc()
+	return errors.Wrap(processingErr, "failed processing SQS message (it will return to queue after visibility timeout)")
+}
+
+func (p *sqsS3EventProcessor) keepalive(ctx context.Context, log *logp.Logger, wg *sync.WaitGroup, msg *sqs.Message) {
+	defer wg.Done()
+
+	t := time.NewTicker(p.sqsVisibilityTimeout / 2)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			log.Debugw("Extending SQS message visibility timeout.",
+				"visibility_timeout", p.sqsVisibilityTimeout,
+				"expires_at", time.Now().UTC().Add(p.sqsVisibilityTimeout))
+			p.metrics.sqsVisibilityTimeoutExtensionsTotal.Inc()
+
+			// Renew visibility.
+			if err := p.sqs.ChangeMessageVisibility(ctx, msg, p.sqsVisibilityTimeout); err != nil {
+				log.Warn("Failed to extend message visibility timeout.", "error", err)
+			}
+		}
+	}
+}
+
+func (p *sqsS3EventProcessor) getS3Notifications(body string) ([]s3EventV2, error) {
+	// NOTE: If AWS introduces a V3 schema this will need updated to handle that schema.
+	var events s3EventsV2
+	dec := json.NewDecoder(strings.NewReader(body))
+	if err := dec.Decode(&events); err != nil {
+		p.log.Debugw("Invalid SQS message body.", "sqs_message_body", body)
+		return nil, fmt.Errorf("failed to decode SQS message body as an S3 notification: %w", err)
+	}
+
+	var out []s3EventV2
+	for _, record := range events.Records {
+		if !p.isObjectCreatedEvents(record) {
+			p.warnOnce.Do(func() {
+				p.log.Warnf("Received S3 notification for %q event type, but "+
+					"only 'ObjectCreated:*' types are handled. It is recommended "+
+					"that you update the S3 Event Notification configuration to "+
+					"only include ObjectCreated event types to save resources.",
+					record.EventName)
+			})
+			continue
+		}
+
+		// Unescape s3 key name. For example, convert "%3D" back to "=".
+		key, err := url.QueryUnescape(record.S3.Object.Key)
+		if err != nil {
+			return nil, fmt.Errorf("url unescape failed for '%v': %w", record.S3.Object.Key, err)
+		}
+		record.S3.Object.Key = key
+
+		out = append(out, record)
+	}
+
+	return out, nil
+}
+
+func (_ *sqsS3EventProcessor) isObjectCreatedEvents(event s3EventV2) bool {
+	return event.EventSource == "aws:s3" && strings.HasPrefix(event.EventName, "ObjectCreated:")
+}
+
+func (p *sqsS3EventProcessor) processS3Events(ctx context.Context, log *logp.Logger, body string) error {
+	s3Events, err := p.getS3Notifications(body)
+	if err != nil {
+		if errors.Is(err, context.Canceled) {
+			// Messages that are in-flight at shutdown should be returned to SQS.
+			return err
+		}
+		return &nonRetryableError{err}
+	}
+	log.Debugf("SQS message contained %d S3 event notifications.", len(s3Events))
+	defer log.Debug("End processing SQS S3 event notifications.")
+
+	// Wait for all events to be ACKed before proceeding.
+	acker := newEventACKTracker(ctx)
+	defer acker.Wait()
```

Comment:
actually one doubt here: do we really want to wait for all the events to be acked for the worker to shutdown?

Reply:
Yes. S3 event notifications for `ObjectCreated` sometimes contain information for more than object. This function returns after all the events generated from each of those objects (usually it's just 1 object) has been ACKed. Then the caller of this function can stop updating the SQS visibility timeout for this one SQS message and decide based on whether an error occurred while processing the S3 object if it should delete the SQS message or try to process it again.

Reply:
Yes, I understand the reason for waiting in the PR, but I was wondering if we should wait for ACK for actually handling the SQS message.
The current implementation doesn't have this: in theory ACK could be blocked for a certain amount of time (network issue reaching elasticsearch, for example), filebeat queue should be able to later catch up. While here we will block starting new workers if I got it right.

Reply:
The reason for waiting until all ACKs are received before deleting the SQS message (and freeing the worker) is that the queued events are non-persistent by default. If we delete the SQS message before receiving all ACKs then Filebeat is vulnerable to data loss if the process is killed before the queue is emptied (events are ACKed).

By waiting until all ACKs are received we ensure that another instance of Filebeat will process the same message again if this process is killed before all events are ACKed.

Reply:
I agree we should not delete the SQS message before receiving all ACKs, but by making the worker wait for the ACKs we prevent it from filling the Filebeat queue in cases of backpressure. If the ACK wait / keepalive / delete message handling was done somewhere else (in a goroutine) that would free the worker for keep processing other SQS messages and filling the queue.

Reply:
We could write it that way.

The current implementation is derived from having a single tuning knob of `max_number_of_messages` that controls the number of inflight messages. If we change the model then we'd add another option that controls the number of workers independently of the `max_number_of_messages`.

I'm hesitant to change it now without doing some testing to see if that is actually a problem. I think that if you size the `max_number_of_messages` parameter appropriately based on the files you're processing and your `queue.mem.events` size then you should be able to keep the internal memory queue full in this model without much of a problem.